src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,211 @@
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+// The operation descriptor is not needed as long as the backend does not have an equivalent type and does not support external workspace.
+using spmv_descr = void *;
+
+void init_spmv_descr(sycl::queue & /*queue*/, oneapi::mkl::sparse::spmv_descr_t *p_spmv_descr) {
+    *p_spmv_descr = nullptr;
+}
+
+sycl::event release_spmv_descr(sycl::queue &queue, oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/,
+                               const std::vector<sycl::event> &dependencies) {
+    return detail::collapse_dependencies(queue, dependencies);
+}
+
+void check_valid_spmv(const std::string function_name, sycl::queue &queue,
+                      oneapi::mkl::transpose opA, oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t y_handle, const void *alpha,
+                      const void *beta) {
+    THROW_IF_NULLPTR(function_name, A_handle);
+    THROW_IF_NULLPTR(function_name, x_handle);
+    THROW_IF_NULLPTR(function_name, y_handle);
+
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    detail::check_all_containers_compatible(function_name, internal_A_handle, x_handle, y_handle);
+    if (internal_A_handle->all_use_buffer()) {
+        detail::check_ptr_is_host_accessible("spmv", "alpha", queue, alpha);
+        detail::check_ptr_is_host_accessible("spmv", "beta", queue, beta);
+    }
+    if (A_view.type_view == oneapi::mkl::sparse::matrix_descr::diagonal) {
+        throw mkl::invalid_argument("sparse_blas", function_name,
+                                    "Matrix view's type cannot be diagonal.");
+    }
+
+    if (A_view.type_view != oneapi::mkl::sparse::matrix_descr::triangular &&
+        A_view.diag_view == oneapi::mkl::diag::unit) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "`unit` diag_view can only be used with a triangular type_view.");
+    }
+
+    if ((A_view.type_view == oneapi::mkl::sparse::matrix_descr::symmetric ||
+         A_view.type_view == oneapi::mkl::sparse::matrix_descr::hermitian) &&
+        opA == oneapi::mkl::transpose::conjtrans) {
```

Comment:
Related to this comment https://github.com/oneapi-src/oneMKL/pull/500#discussion_r1629796550, I was wondering if it makes sense to allow `matrix_descr::symmetric` for complex types or `matrix_descr::hermitian` and `transpose::conjtrans` for real types. My understanding is that they make sense mathematically speaking but are probably not useful and often assumed to be restricted to real (resp. complex) types.
The Intel oneMKL product [spec ](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-1/data-types.html)seems to restrict `conjtrans` to complex type but the oneMKL SYCL interface [one ](file://wsl.localhost/Ubuntu/home/rbiessy/prog/oneAPI-spec/source/elements/oneMKL/build/html/architecture/architecture.html)is a bit more vague. I think it's mostly a question of what would be more useful to the user: assume that there was one mistake or trust that the user understands conjugating a real type is a no-op. What do you think @gajanan-choudhary ?

Reply:
> if it makes sense to allow matrix_descr::symmetric for complex types

Symmetric + complex is definitely needed as the operation/output is mathematically different from Hermitian + complex.

> Intel oneMKL product [spec ](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-1/data-types.html)seems to restrict conjtrans to complex type

I was honestly not aware we had documented `conjtrans` that way. I'll get back to you after confirming what they are doing. For real types + hermitian/conjtrans, On the Intel oneMKL sparse BLAS side, we either already dispatch to or will eventually dispatch to symmetric/trans code-paths as applicable if the user calls hermitian/conjtrans on real types. This is particularly true for the C/Fortran CPU side.

Although a typical user would use a specific FP type in their application, we do have to care about HPC libraries such as PETSc or Kokkos that support multiple data types and operations. Our approach has been to generally support all combinations regardless of whether one is mathematically different from another or not, and deal with correct dispatching on our side.

**IF** BLAS/LAPACK support real + conjtrans/symmetric cases (to be confirmed), then my suggestion would be to directly forward the args to the backend unless the backend. If the backend does not handle the case correctly, a top level check on oneMKL Interfaces side for real data type would be needed to replace `hermitian` with `symmetric` and `conjtrans` replaced with `trans`.

Reply:
Thanks for checking.
Currently I am not testing `symmetric` with complex types not `hermitian` with real types. From what you're saying it sounds like it should be tested. I was mostly wondering whether I should add some error checking. I agree it makes sense to forward the values to the backend and add a top level check if needed. I'll wait for your confirmation.

Reply:
I went with the solution of adding more tests for symmetric with complex types and hermitian and conjtrans with real types in https://github.com/oneapi-src/oneMKL/pull/500/commits/82566e5026444717991adffbd1c809989ae79aea

Reply:
Both LAPACK and BLAS teams said they support real + conjtrans cases to just do real + trans as it is the least surprising behavior for the users. Thanks for the change.

Reply:
Unresolving this comment https://github.com/oneapi-src/oneMKL/pull/500#discussion_r1628122571.

Reply:
These configurations are throwing `unimplemented` now, see https://github.com/oneapi-src/oneMKL/pull/500#discussion_r1731636126

Reply:
Okay, this is fine now. Thanks for the change.